Header and Footer Styling

sprad opened the pull request:
## Description of the change

Adding in header and footer stylings for the public dashboard

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

> Closes #7 

## Checklists

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

sprad commented:
This is just the Header styles, but before I get too far along I want to make sure I'm meeting the spirit and conventions of this new project.  Also, I wanted to ask a few questions which I'll post as review comments.

sprad commented on src/branding-bar/BrandingBar.js:
```diff
@@ -1,8 +1,62 @@
+  margin-left: 16px;
+`;
+
+const BrandingBarLinkContainer = styled.div``;
```
Are we roughly trying to avoid using HTML elements in our render functions and instead trying to name all of our markup elements even if their styles are otherwise empty like this one?

jessex replied:
That does appear to be our convention, yes. Let's keep following it.

sprad commented on src/branding-bar/BrandingBar.js:
```diff
@@ -1,8 +1,62 @@
+const BrandingBarContainer = styled.header`
+  display: flex;
+  justify-content: space-between;
+  margin-bottom: 50px;
```
Honestly, I just hard-coded this value to eyeball match the mocks (it's actually still too small).  At any rate, what would you think about me adding a `spacing` section to the theme to allow us to use consistent spacing throughout the app?  Something like what [Tailwind CSS does](https://github.com/tailwindcss/tailwindcss/blob/master/stubs/defaultConfig.stub.js#L132-L152).  TBD what the actual values would be.

sprad replied:
I've added a proposed solution in the [latest commit](https://github.com/Recidiviz/public-dashboard/pull/22).  If we don't link it, I can just yank the commit.

sprad replied:
And... the correct link:  https://github.com/Recidiviz/public-dashboard/pull/20/commits/c3f3ff0f83880e8e20b515e9101034fa1d6d0ffd

macfarlandian replied:
i am not a huge fan of this approach honestly ... i kind of just philosophically don't actually believe that "consistent spacing" is a thing? In practice I found this feature of Tailwind to be more trouble than it was worth on the COVID site ... the sizing numbers correspond to somewhat arbitrary rem values so I would mostly find myself choosing spacing numbers at semi-random and eyeballing until it looked OK. I don't really see how that's any better than eyeballing pixel values or getting them from Figma

sprad replied:
No worries, I can revert the commit.  I'll still attempt to use semi-consistent spacing values, but we'll see how long that holds up and how well the design is supported (or not supported) in doing so.

sprad commented on src/controls/shared.js:
```diff
@@ -1,17 +1,14 @@
 import styled, { css } from "styled-components";
 
-const PILL_HEIGHT = 32;
+import { PillContainer, PillValue } from "../pill";
```
Since all of these controls were "pills" I created a new `Pill` component so that I could reuse these exact styles for the pill links in the Header.

macfarlandian replied:
hell yeah

sprad commented:
**Screenshot**

<img width="1680" alt="Screen Shot 2020-07-08 at 6 13 25 PM" src="https://user-images.githubusercontent.com/25503/86975771-08a39b00-c147-11ea-9bf2-d0a8ef97887e.png">

sprad commented on src/branding-bar/BrandingBar.js:
```diff
@@ -1,8 +1,62 @@
+    <BrandingBarContainer>
+      <BrandingBarHeader>
+        <Logo alt="Recidiviz" src={LogoIconSrc} />
+        <BrandingBarTitle>North Dakota Corrections</BrandingBarTitle>
```
I've hard-coded "North Dakota Corrections" for now, but would assume we'd like to use some sort of `TenantContext` in the future?

jessex replied:
In the future, absolutely. For now, we want to basically hard-code ND values but not make major design choices that might box us into a single-tenant approach without major effort.

sprad commented on src/branding-bar/BrandingBar.js:
```diff
@@ -1,8 +1,62 @@
+      <BrandingBarLinkContainer>
+        <BrandingBarLinksList>
+          <BrandingBarLink>
+            <LinkPill href="#">Share</LinkPill>
```
These pill-based links look fine on the Desktop, but look pretty janky on mobile due to how much space they take up.  Do we have any thoughts on how we'd like to approach mobile treatments for this app?  Presumably using media queries, but more specifically do we know if styled-components natively jives well with mobile development?

jessex replied:
We need mobile design to be solid at launch since overall polish at launch is going to be important. We should ask Humphrey if there are mobile-first designs anywhere, or if he and Juan have specific adaptations for any elements that may not be mobile-friendly.

As for your last question, I can't speak to that myself. @macfarlandian?

sprad replied:
A proposal:  https://github.com/Recidiviz/public-dashboard/pull/22

macfarlandian replied:
I gestured vaguely at mobile support when getting the bones of the site in place, so the overall layout should be more or less responsive, though I will admit that I don't believe the map chart I created is all the way there, and it is in general not very polished. (I have focused thus far on trying to make things shrink to fit the screen and not so much on whether they actually look great when they do so – in part because there are no small-screen designs that have been shared out yet. That said, I have yet to see anything that looks truly unworkable on mobile.)

I have been trying to avoid media queries except where they are absolutely necessary, in favor of flexible layouts – e.g. setting minimum sizes for the smallest supported screens and letting things grow and wrap as needed with flexbox rules or percentage rules, as applicable. But for something like this that isn't really structural, I think using media queries is probably appropriate. Styled Components [supports them](https://styled-components.com/docs/faqs#can-i-nest-rules) as part of its nesting syntax, so no problems there.

sprad replied:
Ok, cool.  Let me re-work the footer away from the [media query proposal](https://github.com/Recidiviz/public-dashboard/pull/22) and instead try to get it working with only flexbox rules

sprad commented:
**Footer Screenshot**

<img width="1680" alt="Screen Shot 2020-07-08 at 10 42 11 PM" src="https://user-images.githubusercontent.com/25503/86991163-60083200-c16c-11ea-8e26-21524af97a0c.png">

sprad commented on src/branding-bar/BrandingBar.js:
```diff
@@ -1,8 +1,62 @@
+const BrandingBarContainer = styled.header`
+  display: flex;
+  justify-content: space-between;
+  margin-bottom: 75px;
```
> Re-adding this comment since it got blown away when I pushed my latest changes

Honestly, I just hard-coded this margin value to eyeball match the mocks. What would you think about adding a spacing section to the theme to allow us to use consistent spacing throughout the app? Something like what [Tailwind CSS](https://github.com/tailwindcss/tailwindcss/blob/master/stubs/defaultConfig.stub.js#L132-L152) does. TBD what the actual values would be.

jessex replied:
That seems reasonable to me. Would like @macfarlandian to weigh in since I can't think of what the downside would be.

sprad commented on src/constants.js:
```diff
@@ -24,23 +28,30 @@ const displayMedium = `500 ${displayFontSize}/${displayLineHeight} ${displayFont
 
 const brightGreen = "#25b894";
 const darkGreen = "#005450";
+const darkerGreen = "#00413E";
```
I've seen the light, lighter, lightest, dark, darker, darkest pattern used elsewhere.  It has its obvious limitations so I'm open to suggestions for improvement if we don't want to go down this path.

macfarlandian replied:
this is totally fine, i don't see the point in being fussy about this unless we are actually being given an explicit palette of named colors by Juan (which we are not) ... I think that in practice we should try to use the Theme object to define functional colors as much as possible to minimize the number of references to color names, since I would expect them to drift and become more nonsensical over time. 

Plus one of the things coming down the pike is co-branding, meaning we may have to swap out many of these functional colors per state ... so we should get in the habit of not using explicit color names outside the theme.

sprad replied:
Cool, cool.  I'll change this theme value to `footerBackground: darkerGreen;` to make it functional.

sprad commented on src/site-layout/SiteLayout.js:
```diff
@@ -2,13 +2,18 @@ import PropTypes from "prop-types";
+  width: 100%;
+`;
+
+const BodyWrapper = styled.div`
```
The footer needs to extend the width of the screen so I had to push `SiteContainer` up a level since it was previously confined to a width of `1144px`

jessex commented on src/footer/Footer.js:
```diff
@@ -0,0 +1,79 @@
+          </RecidivizBrandingContainer>
+          <FooterLegalContent>
+            <FooterLegalese>
+              &copy; 2019 Recidiviz. All Rights Reserved.
```
2020

sprad replied:
Someone's living in the past...

macfarlandian commented on src/constants.js:
```diff
@@ -24,23 +28,30 @@ const displayMedium = `500 ${displayFontSize}/${displayLineHeight} ${displayFont
     controlLabel: "#3D4045",
-    controlValue: "#5A6575",
+    controlValue: darkGray,
+    footer: hexAlpha("#E5ECEC", 0.6),
```
confession time: I have been converting almost all of the transparencies in the Figma mocks to their RGB equivalents ... unless something has a functional reason to be transparent. I guess I am not super committed to that as an ongoing strategy but it made sense to me at the time? anyway I mention this because I think we should try to at least be consistent one way or another. might be something we should check with Juan about actually

sprad replied:
Ok, I posted the question to Juan but in the meantime will just convert to the RGB equivalent and remove the `hex-alpha` dependency

macfarlandian commented on src/footer/Footer.js:
```diff
@@ -0,0 +1,79 @@
+          </RecidivizBrandingContainer>
+          <FooterLegalContent>
+            <FooterLegalese>
+              &copy; 2020 Recidiviz. All Rights Reserved.
```
maybe overkill but like .... should we just set the year dynamically from a `new Date()`?

sprad replied:
Good idea 👍

macfarlandian commented on src/pill/LinkPill.js:
```diff
@@ -0,0 +1,30 @@
+import { PillContainer, PillValue } from "./shared";
+
+const LinkPillContainer = styled(PillContainer)`
+  a {
+    text-decoration: none;
+  }
```
I think you could just move this to the `LinkPillLink` styles?

sprad replied:
Yup.  Previously I was using an `a` tag directly in the `return` statement below, but when I changed it into the `LinkPillLink` styled component I forgot to move this over.  Great catch 👀

sprad commented:
Ok, I believe I've addressed the remaining PR feedback in the [latest commit](https://github.com/Recidiviz/public-dashboard/pull/20/commits/d99c1caa6690fd3e75f8501c35be46364000a1cc).  A couple of notes:

- Juan says it's fine if we use RGB values directly in our CSS as long as we don't mind doing the conversion.
- Juan also mentioned that he does all of his spacing in multiples of 8 so to the extent it makes sense, we should likely follow suit.
- I've got the footer to be responsive by only using flexbox rules.  It lacks the polish I was able to achieve using media queries in the [Mobile Styles Proposal PR](https://github.com/Recidiviz/public-dashboard/pull/22), but it does grow and shrink dynamically as the screen size changes.  Fit and finish can come later when we're provided with designs for mobile devices.
